### ISSUE TITLE
Encoding lost upon concatenation

### ISSUE THREAD

**Issue by gerritholl** [pydata/xarray#1297]
When using `xarray.concat`, attributes are retained, but encoding information is not.  I believe encoding information should be retained, at least optionally.

```
In [64]: da = xarray.DataArray([1, 2, 3, 2, 1])

In [65]: da.attrs.update(foo="bar")

In [66]: da.encoding.update(complevel=5)

In [67]: da2 = xarray.concat((da, da), dim="new")

In [68]: print(da2.attrs)
OrderedDict([('foo', 'bar')])

In [69]: print(da2.encoding)
{}
```

**Comment by gerritholl** [comment]
This is more serious when we are concatenating datasets, because then the encoding is lost *for each containing data-array*…

**Comment by shoyer** [comment]
I guess I'm not opposed to this *per se*, but preserving `encoding` in operations is even harder than `attrs` because it's possible encoding to no longer be valid. So currently most xarray operation simply delete encoding. That said, I can see some reasons for keeping it when concatenating (e.g., because `concat` is a fundamental part of `open_mfdataset`).

**Comment by gerritholl** [comment]
Mine retains it always upon concatenation, but if you prefer we could add an argument `keep_encoding` in analogy with `keep_attrs`.  In that case we'd want to add it wherever we have `keep_attrs`.

**Comment by shoyer** [comment]
> Mine retains it always upon concatenation, but if you prefer we could add an argument keep_encoding in analogy with keep_attrs. In that case we'd want to add it wherever we have keep_attrs.

No, I'm happy with your fix here. I'd rather keep `encoding` as hidden away from the user facing API as possible, because it's only relevant to a fraction of users (those reading and writing netCDF files).

**Comment by jhamman** [comment]
closed via #1297
